Devel/analysis v3

stephlazar opened the pull request:
Major changes to the analysis_v3 framework. 

Changes are only related to analysis_v3 files, and to RB measurements which use this analysis.
(There is one change in UHFQC_multi_detector where correlated is now default False)

I added the demo notebook I presented in April during the software meeting on analysis_v3.
This is in PycQED_py3/examples/new_features_demos

Commit:
Improves the framework by making the data extraction modular as well:

   * adds extract_data_hdf, add_measured_data_hdf, get_timestamps to pipeline_analysis.py.
   * removes the code to add measured data from get_params_from_hdf_file
   * keeps PipelineDataAnalysis and PipelineDataAnalysis_multi_timestamp for now
   * renames add_measured_data to add_measured_data_dict

Adds RamseyAnalysis class to data_processing.py. Not fully implemented yet.
Renames get_measobj_properties to get_measurement_properties.
Creates the convention to use ana_res_dict as key to save analysis results in data_dict.
Adds save_analysis_results to saving.py and calls it at the end of every analyze_fit_results method.

Adds support for plot_params and params_to_print to all prepare fit functions and to prepare_fit_plot_dicts.

Commit:
Merge branch 'enh/analysis_v3_upgrades' into devel/analysis_v3

Commit:
Finishes RamseyAnalysis implementation.
Adds get_data_file_from_timestamp to pipeline_analysis.py.
Renames the input parameters update_key, append_key, replace_key with update_value, append_value,
   replace_value in add_param and add_measured_data_hdf.
Adds replace_value to add_param.
Modifies get_param to first check whether param is found before splitting it and looking in folders.
Modifies get_default_plot_params to PRX style figures.
Adds the global variables prx_single_column_width, prx_two_column_width, FIGURE_WIDTH_1COL, FIGURE_WIDTH_2COL and
   the functions default_figure_height and add_letter_to_subplots to plotting.py.

Commit:
Fixes the add_letter_to_subplots to check whether there are several columns.

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Adds ramsey_analysis.py and removes RamseyAnalysis class from data_processing.py.
Improves processing_pipeline.py:

 - creates RawPipeline class and create_pipeline function
 - makes check_keys_mobjn a global function
 - removes dependence on the function 'add_'+node_name+'_node' but keeps these functions in the module for now
 - modifies check_keys_mobjn to have the keys_out_container always be the measurement object.

Adds pop_param that pops from dictionaries in addition to get_params which gets from dictionaries.
Improves plotting aspect based on a ramsey measurement.

Commit:
Adds conversion to list if meas_obj_names is string in check_keys_mobjn.
Improves docstring for check_keys_mobjn.

Commit:
Removes the addition of ana_res_dict in ramsey_analysis. All qubit parameters, including the ones
   calculated from analyzing the fit results are added to the qubit dictionary on data_dict.
Removes meas_object_name from key names since they are already included in the keys_in.
Replaces get_sweep_points_from_timestamp with get_param_from_metadata_group to make it more general.
Adds default_figure_title to plotting.py.
Cleans up prepare_1d_plot_dicts and prepare_cal_states_plot_dicts.

Commit:
Improves add_letter_to_subplots.
Cleans up prepare_1d_raw_data_plot_dicts, prepare_2d_plot_dicts, prepare_2d_raw_data_plot_dicts.

Commit:
Major changes to processing_pipeline.py

 * all processing pipelines start out as raw:
   - removes ProcessingPipeline and renames RawPipeline to ProcessingPipeline.
   - add the code from the create_pipeline function to the __call__ method of the new ProcessingPipeline
   - makes check_keys_mobjn a method of the ProcessingPipeline class
 * upgrades check_keys_mobjn to work when meas_object_names is a list of objects names that need to processed
individually, and when keys_in is a list of the form ['previous mobjn.node_name']
 * adds detailed docstring at the beginning of the module

Adds is_string_in to helper_functions.py.

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Adds Save class in saving.py
 - saves figures and creates another HDF5 file for storing data_dict entries
 - removes save_analysis_results
 - makes save_figures and save_fit_results methods of Save.
 - Save is called at the end of process_pipeline
 - removes all other calls the saving.py throughout analysis_v3.

Adds option to specify error_message in add_param and pop_param.
Renames data_dict key 'figs' to 'figures'.

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Adds randomized_benchmarking_analysis.py to analysis_v3 and removes the SingleQubitRBAnalysis class from data_processing.py.

Changes in ProcessingPipeline
 - splits check_keys_mobjn into 3 separate methods; adapts __call__ accordingly
 - keeps the original check_keys_mobjn method as commented out code for now

Fixes bug in analysis_toolbox.py get_timestamps_in_range, which would fail is the day folder contained folders whose first
 6 characters could not be converted to int, or if the day folder contained extra files in addition to folders.

Closes all figures before storing them in data_dict.

Adds option to choose what to save in saving.py Save.

Renames get_channel_names_from_timestamp to get_value_names_from_timestamp in helper_functions.py.
Adds get_reset_reps_from_data_dict to helper_functions.py.

Sets default value of input parameter data_dict in pipeline_analysis.py get_timestamps to None instead of OrderedDict().

Fixes bug in multi_qubit_module.py get_meas_obj_value_names_map that appended extra channels like correlation with the
 same entry for both key and value.
Sets the default value of correlation to False in UHFQC_multi_detector.

Commit:
Adds init_anav3.py.

Improves creation of search_modules in pipeline_analysis.py.
 - search_modules is now a set
 - search_modules is populated with the functions in module upon importing of said module

Improves ProcessingPipeline to create a separate dict for each measured object in meas_obj_names if the node.
 Hence, support for multiple measured objects is handled by this class.

Removes PipelineDataAnalysis and PipelineDataAnalysis_multi_timestamp from pipeline_analysis.py.

Improves process_pipeline to append the processing_pipeline passed as input parameter to the corresponding
 key in data_dict. Also avoids having the same node twice.

Improves add_letter_to_subplots and title placement in plotting.py. Removes presentation_mode plotting in plot.

Commit:
Improves aspects of plots
 - better positioning of titles
 - better positioning of subplot letters

Fixes bugs in rb_analysis related to leakage textboxes.

Adds filename as input parameter to Save class.

Commit:
Improves docstring of add_letter_to_subplots.

Commit:
Adds rabi_analysis.py and removes RabiAnalysis class from data_processing.py.

Fixes bug in get_param and pop_param where the value was not properly extract because the
 default value for the initial try to extract was set to default_value. Also improves code
 to avoid the elementwise comparison failed warning when value is an array or list.

Changes in plotting.py:
 - Improves default_figure_title to add measured object name.
 - Fixes color to red in prepare_fit_plot_dicts.
 - adds fig.align_ylabels()

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Adds support for auto in StateTomographyAnalysis.

Renames threshold_data to classify_data and adds a simpler threshold_data function.
Adds probability_table, measurement_operators_and_results, and calibration_point_means_and_channel_covariations
 to data_processing.py. (In development)
Improves get_param_from_metadata_group.
Adds get_observables, observable_product, and convert_channel_names_to_index to helper_functions.py
Fixes bugs in ProcessingPipeline for joint processing.
Renames the methods in rabi_analysis.py, ramsey_analysis.py, randomized_benchmarking_analysis.py.

Adds extraction of cal states indices based on preparation params in CalibrationPoints.

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

# Conflicts:
#	pycqed/analysis_v3/helper_functions.py

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

# Conflicts:
#	pycqed/analysis_v3/plotting.py
#	pycqed/measurement/calibration_points.py
#	pycqed/measurement/sweep_points.py

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Merge branch 'qudev_master' into devel/analysis_v3

Commit:
Adds memoization function to __call__ of ProcessingPipeline.
Calls processing_pipeline in process_pipeline such that user can pass in unresolved pipeline.
Adds call to get_timestamps in extract_data_hdf.

Commit:
Updates docstring for threshold_data.

Commit:
Updates measure_randomized_benchmarking and measure_two_qubit_randomized_benchmarking
   to the latest analysis_v3 version.

Commit:
Fixes some unwanted diffs with qudev_master.

Commit:
Adds analysis_v3_demo.ipynb.

chellings commented on pycqed/analysis_v3/processing_pipeline.py:
```diff
@@ -1,353 +1,774 @@
+         keys in the data_dict, and most functions expect keys_out
+        - to create the pipeline that will be used by the analysis 
+         framework, the user must call: 
+         ProcessingPipeline_instance(meas_obj_value_names_map), where 
```
this documentation might be outdated (seems that meas_obj_value_names_map is no longer passed at init stage)

stephlazar replied:
It is not completely outdated. The pipeline instance must still be called with meas_obj_value_names map in order to resolve it. However, now I added the option to pass the un-resolved pipeline to the analysis and the analysis resolves it based on meas_obj_value_names_map. So the user does not necessarily need to do it, and I've updated this part of the docstring to say so.

chellings reviewed (approved):
I have not checked the code in detail, but:
- we have seen the demo of analysis_v3, and it makes sense to make this part of the master to encourage people to start getting experience with it
- the RB code has been used successfully on XLD by @nathlacroix and it makes sense to have this in the master (which requires to merge also the analysis_v3 changes)
- the changes do not seem to have an influence on any other current measurements, so it should be save to merge

So, thank you for creating this pull request and for all your work on v3. Only one comment on documentation.

stephlazar commented:
Thanks a lot for looking at it!

Commit:
Updates docstring of ProcessingPipeline.

Commit:
Some improvements in helper_functions.py

Adds closing of data file in get_params_from_hdf_file.
Adds raise_error parameter and option to extract reversed movnm to get_measurement_properties.

stephlazar commented:
I intentionally did not delete the branch.
